app/controllers/users_controller.rb

```diff
@@ -71,4 +126,19 @@ def find_for_authentication(username, provider)
     User.find_for_authentication(username: username, provider: provider) || User.find_or_initialize_by(username: username, provider: provider)
   end
 
+  def passive_shibboleth_url
+    "/Shibboleth.sso/Login?isPassive=true&target=#{uri_component_original_url}"
+  end
+  private :passive_shibboleth_url
+
+  def uri_component_original_url
+    URI::encode(request.original_url, Regexp.new("[^#{URI::PATTERN::UNRESERVED}]"))
+  end
+  private :uri_component_original_url
+
+  def nyu_shibboleth_omniauth_authorize_path
+    user_omniauth_authorize_path(:nyu_shibboleth, institute: current_institute.code, auth_type: :nyu, redirect_to: uri_component_original_url)
+  end
+  private :nyu_shibboleth_omniauth_authorize_path
+
 end
```

Comment:
@hab278 I think you can move these new functions into two modules within `controllers/users`, one called `WhitelistedApplications` and one called `PassiveLogin`, where the functions related to those bits of functionality are separated into the appropriate module.

Then just include the PassiveLogin module in this controller and only have 3 new methods, `check_passive_shibboleth_and_sign_in`, `check_passive_and_sign_client_in` and `after_login_path_for`